### ISSUE TITLE
Add undo/redo functions to TextField.

### ISSUE THREAD

**Issue by JediBurrell** [flutter/flutter#24222]
Android and Chrome OS support full keyboard layouts, providing the ability to use keyboard shortcuts. Regular EditTexts in Android support Ctrl+Z, and Ctrl+Shift+Z for Undo and Redo.

This functionality is currently missing on Flutter's TextField and could provide a worse user experience for Chromebooks than native applications.

**Comment by zoechi** [comment]
also
- <kbd>Ctrl</kbd>-<kbd>x</kbd>
- <kbd>Ctrl</kbd>-<kbd>c</kbd>
- <kbd>Ctrl</kbd>-<kbd>v</kbd>

**Comment by justinmc** [comment]
This is also important for iOS, where originally a shake gesture triggered undo, and now with iOS 13 swiping with 3 fingers can do both undo and redo.

See https://github.com/flutter/flutter/issues/34749 for the new gestures.

**Comment by Timmmm** [comment]
Particularly jarring on Mac where Cmd-V does work, but Cmd-Z does nothing.

**Comment by maks** [comment]
@Hixie is this something that you see being added to the Textfield in the near future or something the sdk will leave to app devs to implement undo/redo for textfields with something like [this package](https://pub.dev/packages/undo) ?

**Comment by Hixie** [comment]
@maks Making this kind of thing work is part of our Web and desktop efforts, but I can't give you a precise timeline. That said, if anyone does want to contribute patches, we are definitely open to such contributions, just join our chat to get started (see the contributor guide for details). FWIW, @justinmc owns this area of the code.

**Comment by venkatd** [comment]
@Hixie @justinmc is there any further reading on how this is the behavior works in a native desktop/web TextField? We're thinking of implementing this within our app and native behavior is not readily obvious to me.

**Comment by justinmc** [comment]
@venkatd I've noticed the same thing, I'm not sure exactly what it's doing under the hood.  There are probably some open source implementations somewhere that we could look at, maybe Chromium.  Or just take a guess for now; it seems like a debounce of some sort to me.

**Comment by venkatd** [comment]
@justinmc we ended up implementing a rough undo/redo implementation internal to our app. Our app now uses  [SuperTextField](https://github.com/superlistapp/super_editor) by @matthew-carroll which is part his rich text project.

From our investigation, we found there was little consistency in how undo/redo was implemented between browsers and applications.

For example: I was surprised to find that many implementations don't debounced. I'm writing this GitHub comment from Chrome in MacOS and:

- If I type "abc", wait even 30 seconds, then type "xyz", undo removes all 6 characters "abcxyz"
- If I type "abcd", hit backspace, then type "xyz", undo takes me back to "abcd" with the d character highlighted

So, at least in this platform, it appears the checkpointing is related to repeating the same action rather than the time that has passed.

That being said, I think a reasonable undo/redo implementation will have a pretty high impact.

We found that, because there is no consistency among implementations, users don't have a strong expectation on how it undo behave as long as it works and isn't too granular. They hit cmd+z until they go back in time to the point they desire and don't think about it.
